### ISSUE TITLE
file naming of cached models for `load_state_dict_from_url`

### ISSUE THREAD

**Issue by faroit** [pytorch/hub#47]
* I am currently using a torch hub entrypoint for an unreleased model which is in a private github repository.  Therefore the entrypoint is manually called using `hubconf.entrypoint()`.

The entrypoint then calls `load_state_dict_from_url` which takes care of the caching of a model by (default) storing downloaded models in `~/.cache/torch/checkpoints/`. 

Now I noticed, that the filename are directly the `model-<sha256>.pth` filestaken from the url. I think it would be nicer if we could add a prefix (such as the entrypoint) to the downloaded checkpoints for better file organization.

maybe something like `load_state_dict_from_url(..., prefix="XYZGAN-")` would work?

Feel free to move this to pytorch core, if this issue is better suited there.

**Comment by ailzhang** [comment]
Hi @faroit , thanks for the proposal! 
You might have also noticed that we use cached folder if there's one in `~/.cache/torch/` and you don't `force_reload=True`, which prefixed checkpoint do you load when you call `torch.hub.load()`? 
Adding a `prefix` kwarg in `torch.hub.load()` might work. Do you know what're the common uses cases for this?

**Comment by faroit** [comment]
> You might have also noticed that we use cached folder if there's one in ~/.cache/torch/ and you don't force_reload=True, which prefixed checkpoint do you load when you call torch.hub.load()?

yes, I saw that but I do not use `torch.hub.load` for now (as it doesn't support private repositories)

> Adding a prefix kwarg in torch.hub.load() might work. Do you know what're the common uses cases for this?

a) what happens if two models with the identical name (and hash) are to be cached?

**Comment by faroit** [comment]
@ailzhang should this issue be moved to the pytorch core repo?
